### PR TITLE
SCT-755  strict mapping

### PR DESCRIPTION
**Pull request by arfathpasha**
- added dynamic strict mapping
- fixed a bug with indexing returning successfully (bulk indexing returning status code 200) when some indexing operations fail

### CODE REVIEW

File: lib/src/kbasesearchengine/search/ElasticIndexingStorage.java [hunk 1]
```diff
@@ -371,14 +374,46 @@ public void indexObjects(
...
+    private void verify(Response response) throws IOException {
+        // convert JSON string to Map
+        String responseJsonStr = EntityUtils.toString(response.getEntity());
+
+        ObjectMapper mapper = new ObjectMapper();
+        final Map<String, Object> responseMap = mapper.readValue(responseJsonStr,
+                new TypeReference<Map<String, Object>>() {});
+
+        // check for and collect response data from index operations that failed
+        List errorItems = new ArrayList();
```
Comment (MrCreosote) [inline lib/src/kbasesearchengine/search/ElasticIndexingStorage.java]: final

Comment (MrCreosote) [inline lib/src/kbasesearchengine/search/ElasticIndexingStorage.java]: Should there be a max size for this list? The exception text could be really big for a large load. Maybe log everything but only throw a portion of the list?

File: lib/src/kbasesearchengine/search/ElasticIndexingStorage.java [hunk 2]
```diff
@@ -371,14 +374,46 @@ public void indexObjects(
...
+        if(responseMap.get("errors").equals(Boolean.TRUE)) {
```
Comment (MrCreosote) [inline lib/src/kbasesearchengine/search/ElasticIndexingStorage.java]: space to distinguish key words from functions all through this part

File: lib/src/kbasesearchengine/search/ElasticIndexingStorage.java [hunk 3]
```diff
@@ -371,14 +374,46 @@ public void indexObjects(
...
+            for(Map item: (List<Map>)responseMap.get("items")) {
+                for(Object itemData: item.values()) {
+                    if (!((Map) itemData).get("status")
+                            .toString().startsWith("20")) {  // status code 200, 201 etc
+                        errorItems.add(itemData);
```
Comment (MrCreosote) [inline lib/src/kbasesearchengine/search/ElasticIndexingStorage.java]: How big is itemData?

File: lib/src/kbasesearchengine/search/ElasticIndexingStorage.java [hunk 4]
```diff
@@ -371,14 +374,46 @@ public void indexObjects(
...
+                    }
+                }
+            }
+        }
+
+        // report response data from failed index operations
+        if(errorItems.size() > 0) {
+            throw new IOException("Indexing errors: " + errorItems.toString());
```
Comment (MrCreosote) [inline lib/src/kbasesearchengine/search/ElasticIndexingStorage.java]: Doesn't seem like there's any tests for this

File: test/src/kbasesearchengine/test/integration/SearchAPIIntegrationTest.java [hunk 5]
```diff
@@ -629,8 +627,52 @@ public void pruneNarrative() throws Exception {
         assertThat("incorrect keyprops", od.getKeyProps(),
                 is(ImmutableMap.of("title", "a title")));
     }
-    
-    /* ****** Auth client tests - to be moved to their own suite *** 
+
+
+    @Test
```
Comment (MrCreosote) [inline test/src/kbasesearchengine/test/integration/SearchAPIIntegrationTest.java]: This seems like it'd be a better fit for a unit test rather than an integration test.

File: lib/src/kbasesearchengine/search/ElasticIndexingStorage.java [hunk 6]
```diff
@@ -23,6 +23,8 @@
 import java.util.function.Function;
 import java.util.stream.Collectors;
 
+import com.fasterxml.jackson.core.type.TypeReference;
```
Comment (MrCreosote) [inline lib/src/kbasesearchengine/search/ElasticIndexingStorage.java:26]: Might want to prevent dynamic index creation as well (not sure about this since it's cluster wide):
```
PUT _cluster/settings
{
 "persistent": {
   "action.auto_create_index" : false
 }
}
```

File: test/src/kbasesearchengine/test/integration/SearchAPIIntegrationTest.java [hunk 7]
```diff
@@ -584,21 +585,18 @@ public void status() throws Exception {
     public void pruneNarrative() throws Exception {
         wsCli1.createWorkspace(new CreateWorkspaceParams()
                 .withWorkspace("narprune"));
-        
+
         final Map<String, Object> data = new HashMap<>();
         data.put("source", "a long string");
```
Comment (MrCreosote) [inline test/src/kbasesearchengine/test/integration/SearchAPIIntegrationTest.java]: Why is this test changing? Seems like the changes defeat the purpose of the test.

File: lib/src/kbasesearchengine/search/ElasticIndexingStorage.java (hunk 1 above)
Comment (arfathpasha) [reply lib/src/kbasesearchengine/search/ElasticIndexingStorage.java]: Fixed.

Comment (arfathpasha) [reply lib/src/kbasesearchengine/search/ElasticIndexingStorage.java]: It won't be larger than the load itself so I am not sure if I should just put some arbitrary number. Logging would be great but at the moment there is no logging in ElasticIndexingStorage. I will create a separate PR for that.

File: lib/src/kbasesearchengine/search/ElasticIndexingStorage.java (hunk 2 above)
Comment (arfathpasha) [reply lib/src/kbasesearchengine/search/ElasticIndexingStorage.java]: Fixed.

File: lib/src/kbasesearchengine/search/ElasticIndexingStorage.java (hunk 3 above)
Comment (arfathpasha) [reply lib/src/kbasesearchengine/search/ElasticIndexingStorage.java]: Not very big. 3 or 4 key value pairs.

File: lib/src/kbasesearchengine/search/ElasticIndexingStorage.java (hunk 4 above)
Comment (arfathpasha) [reply lib/src/kbasesearchengine/search/ElasticIndexingStorage.java]: SearchAPIIntegrationTest.strictMappingFail() covers this case.

File: test/src/kbasesearchengine/test/integration/SearchAPIIntegrationTest.java (hunk 5 above)
Comment (arfathpasha) [reply test/src/kbasesearchengine/test/integration/SearchAPIIntegrationTest.java]: I prefer the integration test because the verification code relies on the ES response. If I mock the ES error response and we change ES version in the future to a version that has a slightly different message, the unit test will silently pass each time and the code may become buggy.

File: lib/src/kbasesearchengine/search/ElasticIndexingStorage.java (hunk 6 above)
Comment (arfathpasha) [reply lib/src/kbasesearchengine/search/ElasticIndexingStorage.java:26]: I'll have to look into this and make a separate PR for it. I have to think about how to do this as a one time operation for the entire index.

File: test/src/kbasesearchengine/test/integration/SearchAPIIntegrationTest.java (hunk 7 above)
Comment (arfathpasha) [reply test/src/kbasesearchengine/test/integration/SearchAPIIntegrationTest.java]: Because it was violating the strict mapping rule.

File: lib/src/kbasesearchengine/search/ElasticIndexingStorage.java (hunk 1 above)
Comment (MrCreosote) [reply lib/src/kbasesearchengine/search/ElasticIndexingStorage.java]: The load size could be up to 100k docs - that's a really really large exception string. 

Logging should be easy - just use slf4j and logback classic (which may already be in the ant file) and do 

```
	private void logInfo(final String format, final Object... params) {
		LoggerFactory.getLogger(getClass()).info(format, params);
	}
	
	private void logErr(final String format, final Object... params) {
		LoggerFactory.getLogger(getClass()).error(format, params);
	}
```

File: lib/src/kbasesearchengine/search/ElasticIndexingStorage.java (hunk 4 above)
Comment (MrCreosote) [reply lib/src/kbasesearchengine/search/ElasticIndexingStorage.java]: Ok, so the test really tests 2 separate things - can you note that in comments in the test?

File: test/src/kbasesearchengine/test/integration/SearchAPIIntegrationTest.java (hunk 5 above)
Comment (MrCreosote) [reply test/src/kbasesearchengine/test/integration/SearchAPIIntegrationTest.java]: Let me rephrase - this seems like it belongs in ElasticIndexingStorageTest (which aren't unit tests, strictly speaking) rather than the API integration test

File: lib/src/kbasesearchengine/search/ElasticIndexingStorage.java (hunk 6 above)
Comment (MrCreosote) [reply lib/src/kbasesearchengine/search/ElasticIndexingStorage.java:26]: Rethinking this - there may be other applications using the same ES install that need dynamic index creation. Maybe just document it in setup that it's preferred but not required.

File: test/src/kbasesearchengine/test/integration/SearchAPIIntegrationTest.java (hunk 7 above)
Comment (MrCreosote) [reply test/src/kbasesearchengine/test/integration/SearchAPIIntegrationTest.java]: Then shouldn't the test should be updated so it doesn't violate that rule but still tests what it's intended to test?

File: lib/src/kbasesearchengine/search/ElasticIndexingStorage.java (hunk 1 above)
Comment (MrCreosote) [reply lib/src/kbasesearchengine/search/ElasticIndexingStorage.java]: That'll be automatically be intercepted by the built in logger in the SearchAPI and will just go to stdout for the workers, I believe.

Comment (arfathpasha) [reply lib/src/kbasesearchengine/search/ElasticIndexingStorage.java]: I have removed the exception and added a statement to log the errors.

File: lib/src/kbasesearchengine/search/ElasticIndexingStorage.java (hunk 4 above)
Comment (arfathpasha) [reply lib/src/kbasesearchengine/search/ElasticIndexingStorage.java]: Not sure I follow. I am only testing one assertion, which has now changed to verifying the log statement and I have added a comment for this.

File: test/src/kbasesearchengine/test/integration/SearchAPIIntegrationTest.java (hunk 5 above)
Comment (arfathpasha) [reply test/src/kbasesearchengine/test/integration/SearchAPIIntegrationTest.java]: Moved to ElasticIndexingStorageTest.

File: lib/src/kbasesearchengine/search/ElasticIndexingStorage.java (hunk 6 above)
Comment (arfathpasha) [reply lib/src/kbasesearchengine/search/ElasticIndexingStorage.java:26]: Added a note in ops.rst documentation in a separate PR

File: test/src/kbasesearchengine/test/integration/SearchAPIIntegrationTest.java (hunk 7 above)
Comment (arfathpasha) [reply test/src/kbasesearchengine/test/integration/SearchAPIIntegrationTest.java]: That's exactly the change that I made.

Comment (scanon) [reply test/src/kbasesearchengine/test/integration/SearchAPIIntegrationTest.java]: I'm confused too. I was guessing the point of this test was to make sure that these other key values are dropped from the indexed object.  Why is it violating the strict mapping rule if they are getting dropped?

**Comment by arfathpasha** [comment]
The "data" in the test is not raw data, but is supposed to simulate parsed data that has been extracted based on the indexing rules. On line 606, you can see that it is being used to create an instance of ParsedObject which is one of the arguments to the indexObjects() method. To make this clear, I have renamed "data" to "parsedData".

```
ImmutableMap.of(new GUID("WS:1/1/1"), new ParsedObject(
                         new ObjectMapper().writeValueAsString(data),
                         data.entrySet().stream().collect(Collectors.toMap(
                                 e -> e.getKey(), e -> Arrays.asList(e.getValue()))))),
                 false);
```

File: test/src/kbasesearchengine/test/integration/SearchAPIIntegrationTest.java (hunk 7 above)
Comment (scanon) [reply test/src/kbasesearchengine/test/integration/SearchAPIIntegrationTest.java]: But this test is meant to test the pruning that happens in the search api.

Here
https://github.com/kbase/KBaseSearchEngine/blob/master/lib/src/kbasesearchengine/main/TemporaryNarrativePruner.java

So the idea, as I understand it, is the index record has these keys and we want to make sure the search api is dropping them.  So excluding them from the start doesn't test the pruner.

I'm still confused as to why the strict mapping is breaking things because those should be valid keys.

Comment (arfathpasha) [reply test/src/kbasesearchengine/test/integration/SearchAPIIntegrationTest.java]: I totally missed the documentation on top of the test about this test belonging to the narrative pruner. I have put back the keys that get pruned to show that the pruning functionality is working. To answer your question about why the strict mapping is breaking this test, it is because the keys in the parsed object need to be backed up with indexing rules as the indexing rules are used to create the mapping. No rule for a key, implies no mapping for that key and this causes the indexing to fail. So now, all I have done is add indexing rules for each of the keys.

**Comment by scanon** [comment]
Looks good.
